README
NAME
    CSS::SpriteBuilder - CSS sprite builder.

SYNOPSIS
        use CSS::SpriteBuilder

        my $builder = CSS::SpriteBuilder->new( [%args] );
        $builder->build(
            sprites => [{
                file   => 'sample_sprite_%d.png',
                images => [
                    { file => 'small/Add.png', [ %options ] },
                ],
                [ %options ],
            }],
        );

        $build->write_css('sprite.css');

        Or

        $builder->build(config => 'config.xml');

        $build->write_css('sprite.css');

DESCRIPTION
    This module generate CSS sprites with one of these modules:
    Image::Magick or GD.

    It has many useful settings and can be used for sites with complex
    structure.

METHODS
  new(<%args>)
    my $builder = CSS::SpriteBuilder->new(<%args>);

    Create instance.

    Valid arguments are:

    *   source_dir [ = undef ]

        Specify custom location for source images.

    *   output_dir [ = undef ]

        Specify custom location for generated images.

    *   image_quality 0..100 [ = 90 ]

        Specify image quality for generated images (for JPEG only).

    *   max_image_size [ = 65536 ]

        Specify max size of images that will be used to generate a sprite.

    *   max_image_width [ = 2000 ]

        Specify max width of images that will be used to generate a sprite.

    *   max_image_height [ = 2000 ]

        Specify max height of images that will be used to generate a sprite.

    *   max_sprite_width [ = 2000 ]

        Specify max width of sprite. When sprite has no free space, than
        creates a new sprite with addition of a suffix to the sprite name.
        Opera 9.0 and below have a bug which affects CSS background offsets
        less than -2042px. All values less than this are treated as -2042px
        exactly.

    *   max_sprite_height [ = 2000 ]

        Specify max height of sprite.

    *   margin [ = 0 ]

        Add margin to each image.

    *   transparent_color [ = undef ]

        Set transparent color for image, for example: 'white', 'black', ....

    *   is_background [ = 0 ]

        If is_background flag is '0' will be generated CSS rule such as:
        'width:XXXpx;height:YYYpx;'.

    *   layout [ = 'packed' ]

        Specify layout algorithm (horizontal, vertical or packed).

    *   css_selector_prefix [ = '.spr-' ]

        Specify CSS selector prefix. For example, for an image
        "img/icon/arrow.gif" will be generated selector such as
        ".spr-img-icon-arrow".

    *   css_url_prefix [ = '' ]

        Specify prefix for CSS url. For example: background-image:
        url('css_url_prefixsample_sprite.png')

    *   is_add_timestamp [ = 1 ]

        If parameter set to '1' than timestamp will be added for CSS url.
        For example: background-image: url('sample_sprite.png?12345678')

  build(<%args>)
    Build sprites.

        $builder->build(<%args>);

    This method returning structure like:

        {
            'sample_sprite_1.png' => [
                {
                    'y' => 0,
                    'width' => 32,
                    'selector' => '.spr-small-add',
                    'is_transparent' => 1,
                    'is_background' => 0,
                    'x' => 0,
                    'height' => 32,
                    'image' => 'small/Add.png',
                    'repeat' => 'no'
                },
                ...
            ],
            ...
        }

    Valid arguments are:

    *   sprites

        Specify sprite list.

            $builder->build(
                sprites => [
                    {
                        file               => 'horizontal_%d.png',
                        layout             => 'horizontal',
                        max_sprite_width   => 1000,
                        ...
                        images => [
                            { file => 'small/*.png', is_repeat => 1, ... },
                            { file => 'small/*.gif' },
                            { file => 'small/a.jpg', is_background => 1 },
                        ],
                    },
                    {
                        file                => 'sprite_%d.png',
                        max_sprite_width    => 1000,
                        css_selector_prefix => '#spr-',
                        ...
                        images => [
                            { file => 'small/Add.png', is_repeat => 1, ... },
                        ],
                    },
                ],
            );

    *   config

        Specify XML config filename (it requires XML::LibXML module).

            $builder->build(config => 'config.xml');

        Example of config.xml:

            <root>
                <global_opts>
                    <max_image_size>20000</max_image_size>
                    <layout>packed</layout>
                    <css_selector_prefix>.spr-</css_selector_prefix>
                    <css_url_prefix>/sprite/</css_url_prefix>
                </global_opts>

                <sprites>
                    <sprite file="sprite_%d.png">
                        <image file="small/Add.png"/>
                        <image file="small/Box.png"/>
                        <image file="medium/CD.png"/>
                    </sprite>
                    <sprite file="sprite_x_%d.png" layout="vertical">
                        <image file="small/Brick.png" is_repeat="1"/>
                        <image file="small/Bin_Empty.png"/>
                        <image file="medium/Close.png"/>
                    </sprite>
                    <sprite file="sprite_y_%d.png" layout="horizontal">
                        <image file="small/Pattern.png" is_repeat="1"/>
                        <image file="small/Address_Book.png"/>
                        <image file="medium/Chat.png"/>
                    </sprite>
                </sprites>
            </root>

  write_css([<filename|fh>])
    Write CSS to file. When filename parameter is not specified than this
    method returning a string.

        $builder->write_css('sprite.css');

  write_xml([<filename|fh>])
    Write CSS sprites info structure into XML format (it requires
    XML::LibXML module). When filename parameter is not specified than this
    method returning a string.

        $builder->write_xml('sprite.xml');

    Example of sprite.xml:

        <root>
            <sprite src="sample_sprite.png">
                <image y="0" width="32" selector=".spr-small-add" is_transparent="1" is_background="0" x="0" height="32" repeat="no" image="small/Add.png"/>
            </sprite>
        </root>

  write_html([<filename|fh>])
    Write HTML sample page. When filename parameter is not specified than
    this method returning a string.

  get_sprites_info()
    This method returning structure like:

        {
            'sample_sprite.png' => [
                {
                    'y' => 0,
                    'width' => 32,
                    'selector' => '.spr-small-add',
                    'is_transparent' => 1,
                    'is_background' => 0,
                    'x' => 0,
                    'height' => 32,
                    'image' => 'small/Add.png',
                    'repeat' => 'no'
                },
                ...
            ],
            ...
        }

AUTHOR
        Yuriy Ustushenko, <yoreek@yahoo.com>

COPYRIGHT AND LICENSE
    Copyright (C) 2013 Yuriy Ustushenko

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for CSS::SpriteBuilder.

0.02    2013-08-17
        Replaced "use parent" with "use base".

0.01    2013-08-11
        First version


Makefile.PL
# Makefile.PL for CSS::SpriteBuilder.
#
# This program is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.
# Copyright 2013 Yuriy Ustushenko, all rights reserved.

use 5.008008;
use warnings;
use strict;
use ExtUtils::MakeMaker;

our @MODULES = qw( Image::Magick GD );

my $isOK = 0;
print "This module requires at least one of: ", join(", ", @MODULES), "...\n";
foreach (@MODULES){
    print "Looking for $_ ... ";
    eval "require "."$_";
    if ($@) {
        print "not found\n";
    }
    else {
        print "ok\n";
        $isOK = 1;
        last;
    }
}

our @REQUIRED;
unless ($isOK) {
    print <<EOF;

This module requires you have installed on your system
one of either the Image::Magick or GD Perl modules.

It appears you have none of these.
EOF
    push @REQUIRED, $MODULES[0];
}

WriteMakefile(
    NAME             => 'CSS::SpriteBuilder',
    VERSION_FROM     => 'lib/CSS/SpriteBuilder.pm',
    ABSTRACT_FROM    => 'lib/CSS/SpriteBuilder.pm',
    AUTHOR           => 'Yuriy Ustushenko <yoreek@yahoo.com>',
    LICENSE          => "perl",
    MIN_PERL_VERSION => 5.008008,
    PREREQ_PM     => {
        'File::Spec'   => 0,
        'File::Glob'   => 0,
        'Cwd'          => 0,
        'Scalar::Util' => 0,
        'Test::More'   => 0,
        @REQUIRED,
    },
);


lib/CSS/SpriteBuilder/Sprite.pm
package CSS::SpriteBuilder::Sprite;

use warnings;
use strict;
use File::Glob ":glob";
use File::Spec;
use Cwd ();
use CSS::SpriteBuilder::Constants;
use CSS::SpriteBuilder::SpriteItem;
use base 'CSS::SpriteBuilder::ImageDriver::Auto';

sub max { $_[ $_[0] < $_[1] ] }

sub new {
    my ($class, @args) = @_;

    my $self = $class->SUPER::new(
        source_dir          => undef,
        source_images       => undef,
        target_file         => undef,

        image_quality       => DEF_IMAGE_QUALITY,
        max_image_size      => DEF_MAX_IMAGE_SIZE,
        max_image_width     => DEF_MAX_IMAGE_WIDTH,
        max_image_height    => DEF_MAX_IMAGE_HEIGHT,
        max_sprite_width    => DEF_MAX_SPRITE_WIDTH,
        max_sprite_height   => DEF_MAX_SPRITE_HEIGHT,
        margin              => DEF_MARGIN,
        transparent_color   => undef,
        is_background       => 0,
        layout              => DEF_LAYOUT,
        css_selector_prefix => DEF_CSS_SELECTOR_PREFIX,
        css_url_prefix      => '',
        is_add_timestamp    => 1,

        _sprite_no          => 1,
        _repeated_images    => [],
        _sprites            => {},
        _x                  => 0,
        _y                  => 0,

        @args,
    );

    unless ( grep { $_ eq $self->{layout} } @{ LAYOUT_LIST() } ) {
        die "Invalid 'layout' parameter: $self->{layout}";
    }

    return $self;
}

sub target_file    { sprintf( $_[0]->{target_file}, $_[0]->{_sprite_no} )                }
sub is_transparent { $_[0]->{target_file} =~ /\.png$/i && $_[0]->SUPER::is_transparent() }

sub _init {
    my ($self, $image) = @_;

    die "The 'target_file' parameter is not defined"    unless $self->{target_file};
    die "The 'source_images' parameter is not defined"  unless $self->{source_images};

    $self->SUPER::_init($image);

    $self->{_x} = $self->{_y} = 0;

    return;
}

sub build {
    my ($self) = @_;

    my $images = $self->search_images();
    foreach my $image (@$images) {
        $self->process_image($image);
    }

    $self->write();

    return $self->{_sprites};
}

sub search_images {
    my ($self) = @_;

    my $save_dir;
    if ( $self->{source_dir} ) {
        $save_dir = Cwd::getcwd();
        chdir $self->{source_dir} or die "Cannot chdir to $self->{source_dir}: $!\n";
    }

    my @images = grep {
           ( !$self->{max_image_size}   || -s $_->source_file() <= $self->{max_image_size}   )
        && ( !$self->{max_image_width}  || $_->width()          <= $self->{max_image_width}  )
        && ( !$self->{max_image_height} || $_->height()         <= $self->{max_image_height} )
        && ( $_->width() <= $self->{max_sprite_width} && $_->height <= $self->{max_sprite_height}
            or die "The image resolution is bigger than maximum configured value for '". $_->source_file() ."'"
        )
    } map {
        my $image_opts = $_;
        map {
            CSS::SpriteBuilder::SpriteItem->new(
                %$image_opts,
                source_file => $_,
            )
        } bsd_glob( $image_opts->{file}, File::Glob::GLOB_ERR );
    } @{ $self->{source_images} };

    if ($save_dir) {
        chdir $save_dir or die "Cannot chdir to $save_dir: $!\n";
    }

    return wantarray ? @images: \@images;
}

sub process_image {
    my ($self, $image) = @_;

    $self->allocate_space($image);

    $self->align_image($image);

    $self->save_info($image);

    $self->save_repeated_images($image)
        if $self->{layout} ne PACKED_LAYOUT && $image->{is_repeat};

    $self->post_process($image);
}

sub align_image {
    my ($self, $image) = @_;

    if ( $self->is_blank() ) {
        $self->reset($image);
    }
    else {
        $self->{_x} += $self->{margin} if $self->{_x};

        $self->extent(
            max($self->width(),  $self->{_x} + $image->width() ),
            max($self->height(), $self->{_y} + $image->height()),
        );

        $self->composite( $image, $self->{_x}, $self->{_y} );
    }

    return;
}

sub allocate_space {
    my ($self, $image) = @_;

    return if $self->is_blank();

    my $x_margin = $self->{_x} == 0 ? 0 : $self->{margin};

    if (
        $self->{layout} eq VERTICAL_LAYOUT
        || ($self->{_x} + $image->width() + $x_margin) > $self->{max_sprite_width}
    ) {
        $self->{_x} = 0;
        $self->{_y} = $self->height() + $self->{margin};
    }

    if (
        $self->{_y} > 0 && (
            $self->{layout} eq HORIZONTAL_LAYOUT
            || ($self->{_y} + $image->height() + $self->{margin}) > $self->{max_sprite_height}
        )
    ) {
        $self->write();
    }

    return;
}

sub save_info {
    my ($self, $image) = @_;

    my (undef, undef, $filename) = File::Spec->splitpath( $self->target_file() );
    my $sprite_image = $self->{css_url_prefix} . $filename;

    $sprite_image .= '?' . time() if $self->{is_add_timestamp};

    push @{ $self->{_sprites}{$sprite_image} }, {
        image           => $image->source_file(),
        width           => $image->width(),
        height          => $image->height(),
        x               => $self->{_x} - $image->offset_x(),
        y               => $self->{_y} - $image->offset_y(),
        selector        => $image->get_css_selector( $self->{css_selector_prefix} ),
        is_background   => (
            defined $image->is_background()
                ? $image->is_background()
                : $self->{is_background}
        ) ? 1 : 0,
        is_transparent  => (
            defined $self->{transparent_color}
                ? $self->{transparent_color}
                : $self->is_transparent()
        ) ? 1 : 0,
        repeat          => $self->{layout} ne PACKED_LAYOUT && $image->is_repeat()
            ? $self->{layout} eq HORIZONTAL_LAYOUT ? 'y' : 'x'
            : 'no'
        ,
    };

    return;
}

sub save_repeated_images {
    my ($self, $image) = @_;

    push @{ $self->{_repeated_images} }, {
        image => $image,
        x     => $self->{_x},
        y     => $self->{_y},
    };

    return;
}

sub post_process {
    my ($self, $image) = @_;

    $self->{ $self->{layout} eq VERTICAL_LAYOUT ? '_y' : '_x' } += $image->width;

    return;
}

sub write {
    my ($self) = @_;

    return if $self->is_blank();

    $self->process_repeated_images();

    $self->set_transparent_color( $self->{transparent_color} )
        if $self->{transparent_color};

    $self->set_quality( $self->{image_quality} );

    $self->SUPER::write( $self->target_file() );

    $self->{_sprite_no}++;

    $self->reset();

    return;
}

sub process_repeated_images {
    my ($self) = @_;

    return unless scalar @{ $self->{_repeated_images} };

    my $meter                = $self->{layout} eq VERTICAL_LAYOUT ? 'width' : 'height';
    my $original_sprite_size = $self->$meter();
    my $max_sprite_size      = $self->{ $self->{layout} eq VERTICAL_LAYOUT ? 'max_sprite_width' : 'max_sprite_height' };
    my @images_size          = reverse sort map { $_->{image}->$meter() } @{ $self->{_repeated_images} };
    my $max_image_size       = shift @images_size;

    for (
        my $k = int($original_sprite_size / $max_image_size) + ($original_sprite_size % $max_image_size ? 1 : 0);
        $k * $max_image_size <= $max_sprite_size;
        $k++
    ) {
        my $extented_sprite_size = $k * $max_image_size;
        unless (grep { $extented_sprite_size % $_ } @images_size) {
            $self->extent(
                $meter eq 'width'
                    ? ( $extented_sprite_size, $self->height()        )
                    : ( $self->width()       , $extented_sprite_size  )
            );

            foreach my $image_info (@{ $self->{_repeated_images} }) {
                my $image = $image_info->{image};
                my $size  = $image->$meter();

                while (($image_info->{ $self->{layout} eq VERTICAL_LAYOUT ? 'x' : 'y' } += $size) < $extented_sprite_size) {
                    $self->composite( $image, $image_info->{x}, $image_info->{y} )
                }
            }

            return;
        }
    }

    $self->{_repeated_images} = [];

    die "Can't process repeated images";
}

1;


lib/CSS/SpriteBuilder/Constants.pm
package CSS::SpriteBuilder::Constants;

use warnings;
use strict;

use parent 'Exporter';
our @EXPORT = qw(
    LAYOUT_LIST
    PACKED_LAYOUT
    HORIZONTAL_LAYOUT
    VERTICAL_LAYOUT
    SPRITE_OPTS
    IMAGE_OPTS
    DEF_IMAGE_QUALITY
    DEF_MAX_IMAGE_SIZE
    DEF_MAX_IMAGE_WIDTH
    DEF_MAX_IMAGE_HEIGHT
    DEF_MAX_SPRITE_WIDTH
    DEF_MAX_SPRITE_HEIGHT
    DEF_MARGIN
    DEF_LAYOUT
    DEF_CSS_SELECTOR_PREFIX
);

use constant {
    LAYOUT_LIST       => [qw/ packed horizontal vertical /],
    PACKED_LAYOUT     => 'packed',
    HORIZONTAL_LAYOUT => 'horizontal',
    VERTICAL_LAYOUT   => 'vertical',
};

use constant SPRITE_OPTS => [qw/
    source_dir
    image_quality
    max_image_size
    max_image_width
    max_image_height
    max_sprite_width
    max_sprite_height
    margin
    transparent_color
    is_background
    layout
    css_selector_prefix
    css_url_prefix
    is_add_timestamp
/];

use constant IMAGE_OPTS => [qw/
    is_background
    is_repeat
    css_selector
/];

# default options
use constant {
    DEF_IMAGE_QUALITY       => 90,
    DEF_MAX_IMAGE_SIZE      => 64 * 1024,
    DEF_MAX_IMAGE_WIDTH     => 2000,
    DEF_MAX_IMAGE_HEIGHT    => 2000,
    DEF_MAX_SPRITE_WIDTH    => 2000,
    DEF_MAX_SPRITE_HEIGHT   => 2000,
    DEF_MARGIN              => 0,
    DEF_LAYOUT              => PACKED_LAYOUT,
    DEF_CSS_SELECTOR_PREFIX => '.spr-',
};

1;


lib/CSS/SpriteBuilder/SpriteItem.pm
package CSS::SpriteBuilder::SpriteItem;

use warnings;
use strict;
use base 'CSS::SpriteBuilder::ImageDriver::Auto';

sub new {
    my ($class, @args) = @_;

    my $self = $class->SUPER::new(
        source_file       => undef,
        is_background     => undef,
        is_repeat         => 0,
        offset_x          => 0,
        offset_y          => 0,
        css_selector      => undef,
        @args,
    );

    my $source_file = $self->{source_file}
        or die "The 'source_file' parameter is required";

    $self->read($source_file);

    return $self;
}

sub source_file   { return $_[0]->{source_file   } }
sub is_background { return $_[0]->{is_background } }
sub is_repeat     { return $_[0]->{is_repeat     } }
sub offset_x      { return $_[0]->{offset_x      } }
sub offset_y      { return $_[0]->{offset_y      } }
sub css_selector  { return $_[0]->{css_selector  } }

sub get_css_selector {
    my ($self, $css_selector_prefix) = @_;

    my $selector = $self->{css_selector};
    unless ($selector) {
        # Convert "img/icon/arrow.gif" to ".spr-img-icon-arrow"
        my (undef, $dirs, $file) = File::Spec->splitpath( $self->{source_file} );
        my @dirs = grep { $_ } File::Spec->splitdir($dirs);

        # remove extension
        $file =~ s/\.[^\.]+$//;
        $selector = lc( ($css_selector_prefix || '') . join('-', @dirs, $file) );

        $selector =~ s/[\s_]+/-/g;
    }

    return $selector;
}

1;


lib/CSS/SpriteBuilder.pm
package CSS::SpriteBuilder;

use strict;
use warnings;
use File::Spec;
use Scalar::Util qw(openhandle);
use CSS::SpriteBuilder::Constants;
use CSS::SpriteBuilder::Sprite;

our $VERSION = '0.02';

sub new {
    my ($class, @args) = @_;

    my $self = bless {
        source_dir          => undef,
        output_dir          => undef,

        image_quality       => DEF_IMAGE_QUALITY,
        max_image_size      => DEF_MAX_IMAGE_SIZE,
        max_image_width     => DEF_MAX_IMAGE_WIDTH,
        max_image_height    => DEF_MAX_IMAGE_HEIGHT,
        max_sprite_width    => DEF_MAX_SPRITE_WIDTH,
        max_sprite_height   => DEF_MAX_SPRITE_HEIGHT,
        margin              => DEF_MARGIN,
        transparent_color   => undef,
        is_background       => 0,
        layout              => DEF_LAYOUT,
        css_selector_prefix => DEF_CSS_SELECTOR_PREFIX,
        css_url_prefix      => '',
        is_add_timestamp    => 1,

        _sprites            => {},

        @args,
    }, $class;

    return $self;
}

sub build {
    my ($self, %args) = @_;

    my $sprites;
    if ( my $config = delete $args{config} ) {
        $sprites = $self->_parse_xml_config($config);
    }
    else {
        $sprites = delete $args{sprites};
    }

    foreach my $sprite_opts (@$sprites) {
        my $images      = delete $sprite_opts->{images};
        my $output_file = $self->{output_dir}
            ? File::Spec->catfile( $self->{output_dir}, delete $sprite_opts->{file} )
            : delete $sprite_opts->{file}
        ;

        my $sprite = CSS::SpriteBuilder::Sprite->new(
            ( map { $_ => $self->{$_} } grep { exists $self->{$_} } @{ SPRITE_OPTS() } ),
            %$sprite_opts,
            source_images => $images,
            target_file   => $output_file,
        );

        my $sprites = $sprite->build();
        $self->{_sprites} = {%{ $self->{_sprites} }, %$sprites};
    }

    return $self->{_sprites};
}

sub get_sprites_info {
    my ($self) = @_;
    return $self->{_sprites};
}

sub write_css {
    my ($self, $filename) = @_;

    my ($fh, $str);

    if ($filename) {
        if ( openhandle($filename) ) {
            $fh = $filename;
        }
        else {
            open($fh, '>', $filename) or die "Can't to open file '$filename'!";
        }
    }
    else {
        $str = '';
        open($fh, '>', \$str) or die "Can't to create file handle!";
    }

    while (my ($sprite_image, $images) = each %{ $self->{_sprites} }) {
        my @selectors = map { $_->{selector} } @$images;
        print $fh sprintf(
            "%s{background-image: url('%s') !important;}\n",
            join(',', @selectors),
            $sprite_image,
        );

        foreach my $image_info (@$images) {
            print $fh sprintf(
                "%s{background-position:%dpx %dpx !important;",
                $image_info->{selector},
                -$image_info->{x},
                -$image_info->{y},
            );

            if ( $image_info->{repeat} ne 'no' ) {
                print $fh sprintf(
                    "background-repeat:repeat-%s;",
                    $image_info->{repeat},
                );
            }

            unless ( $image_info->{is_background} ) {
                print $fh sprintf(
                    "width:%dpx;height:%dpx;",
                    $image_info->{width},
                    $image_info->{height},
                );
            }

            print $fh "}\n";
        }
    }

    unless ($filename) {
        close $fh;
        return $str;
    }

    return;
}

sub write_xml {
    my ($self, $filename) = @_;

    eval "require XML::LibXML";
    die "XML::LibXML module is required!" if $@;

    my $dom     = XML::LibXML::Document->new();
    my $root_el = $dom->createElement('root');
    $dom->setDocumentElement($root_el);

    while (my ($sprite_image, $images) = each %{ $self->{_sprites} }) {
        my $sprite_el = $dom->createElement('sprite');
        $root_el->appendChild($sprite_el);
        $sprite_el->setAttribute('src', $sprite_image);

        foreach my $image (@$images) {
            my $image_el = $dom->createElement('image');
            $sprite_el->appendChild($image_el);
            while (my ($key, $value) = each %$image) {
                $image_el->setAttribute($key, $value);
            }
        }
    }

    if ($filename) {
        if ( openhandle($filename) ) {
            return $dom->toFH($filename, 2);
        }
        else {
            return $dom->toFile($filename, 2);
        }
    }

    return $dom->toString(2);
}

sub write_html {
    my ($self, $filename) = @_;

    my ($fh, $str);

    if ($filename) {
        if ( openhandle($filename) ) {
            $fh = $filename;
        }
        else {
            open($fh, '>', $filename) or die "Can't to open file '$filename'!";
        }
    }
    else {
        $str = '';
        open($fh, '>', \$str) or die "Can't to create file handle!";
    }

    print $fh <<HTML;
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
    <title>CSS::SpriteBuilder generated page</title>
    <style type="text/css">
        .container {
            width: 800px;
            margin: 0 auto;
            text-align: center;
            font-size: 12px;
        }
        table {
            border-collapse: collapse;
            width: 100%;
        }
        th, td {
            border: 1px solid;
            text-align: center;
            padding: 8px;
        }
        .spr {
            display: -moz-inline-stack;
            display: inline-block;
            zoom: 1;
            *display: inline;
            background-repeat: no-repeat;
        }
HTML

    $self->write_css($fh);

    print $fh <<HTML;
    </style>
</head>
<body>
    <div class="container">
    <h3>CSS::SpriteBuilder generated page</h3>
    <table>
        <thead>
            <tr>
                <th>Image</th>
                <th>CSS selector</th>
                <th>Sprite</th>
                <th>Pos</th>
            </tr>
        </thead>
        <tbody>
HTML

    while (my ($sprite_image, $images) = each %{ $self->{_sprites} }) {
        for (my $i = 0; $i < scalar @$images; $i++) {
            my $image_info = $images->[$i];
            my $selector   = $image_info->{selector};

            my $rule = $selector =~ /^#/
                ? 'class="spr" id="'. substr($selector, 1) .'"'
                : 'class="spr '. substr($selector, 1) .'"'
            ;

            if ($image_info->{repeat} eq 'x') {
                $rule .= ' style="width:'. $image_info->{width} * 3 .'px;"';
            }
            elsif ($image_info->{repeat} eq 'y') {
                $rule .= ' style="height:'. $image_info->{height} * 3 .'px;"';
            }

            print $fh <<HTML;
            <tr>
                <td>
                    <span $rule>&nbsp;</span><br/>
                    $image_info->{image}<br/>
                    $image_info->{width} x $image_info->{height} (repeat-$image_info->{repeat})
                </td>
                <td>$selector</td>
HTML
            if ($i == 0) {
                my $rowspan = scalar @$images;
                print $fh <<HTML;
                    <td rowspan="$rowspan">
                        <img alt="" src="$sprite_image" /><br/>
                        $sprite_image
                    </td>
HTML
            }

            print $fh <<HTML;
                <td>$image_info->{x} x $image_info->{y}</td>
            </tr>
HTML
        }
    }

    print $fh <<HTML;
        </tbody>
    </table>
    </div>
</body>
</html>
HTML

    unless ($filename) {
        close $fh;
        return $str;
    }

    return;
}

sub _parse_xml_config {
    my ($self, $filename) = @_;

    eval "require XML::LibXML";
    die "XML::LibXML module is required!" if $@;

    my $xml_parser = XML::LibXML->new();

    my $dom;
    if ( openhandle($filename) ) {
        $dom = $xml_parser->parse_fh($filename);
    }
    elsif ( $filename =~ /^</ ) {
        $dom = $xml_parser->parse_string($filename);
    }
    else {
        $dom = $xml_parser->parse_file($filename);
    }

    my %global_opts = (
        ( map { $_             => $self->{$_}       } grep { exists $self->{$_} } @{ SPRITE_OPTS() } ),
        ( map { $_->nodeName() => $_->textContent() } $dom->findnodes("/root/global_opts/*")        ),
    );

    my @sprites;
    foreach my $sprite_node ( $dom->findnodes("/root/sprites/sprite") ) {
        my %sprite_opts = (
            %global_opts,
            map { $_->nodeName() => $_->textContent() } $sprite_node->findnodes("@*"),
        );

        my @images = map +{
            ( map { $_             => $sprite_opts{$_}  } grep { exists $sprite_opts{$_} } @{ IMAGE_OPTS() }   ),
            ( map { $_->nodeName() => $_->textContent() } $_->findnodes("@*")                                  ),
        }, $sprite_node->findnodes('image');

        push @sprites, {
            %sprite_opts,
            images => \@images,
        };
    }

    return wantarray ? @sprites : \@sprites;
}

1;
__END__

=head1 NAME

CSS::SpriteBuilder - CSS sprite builder.

=head1 SYNOPSIS

    use CSS::SpriteBuilder

    my $builder = CSS::SpriteBuilder->new( [%args] );
    $builder->build(
        sprites => [{
            file   => 'sample_sprite_%d.png',
            images => [
                { file => 'small/Add.png', [ %options ] },
            ],
            [ %options ],
        }],
    );

    $build->write_css('sprite.css');

    Or

    $builder->build(config => 'config.xml');

    $build->write_css('sprite.css');

=head1 DESCRIPTION

This module generate CSS sprites with one of these modules: Image::Magick or GD.

It has many useful settings and can be used for sites with complex structure.

=head1 METHODS

=head2 new(<%args>)

my $builder = CSS::SpriteBuilder->new(<%args>);

Create instance.

Valid arguments are:

=over

=item * B<source_dir> [ = undef ]

Specify custom location for source images.

=item * B<output_dir> [ = undef ]

Specify custom location for generated images.

=item * B<image_quality> 0..100 [ = 90 ]

Specify image quality for generated images (for JPEG only).

=item * B<max_image_size> [ = 65536 ]

Specify max size of images that will be used to generate a sprite.

=item * B<max_image_width> [ = 2000 ]

Specify max width of images that will be used to generate a sprite.

=item * B<max_image_height> [ = 2000 ]

Specify max height of images that will be used to generate a sprite.

=item * B<max_sprite_width> [ = 2000 ]

Specify max width of sprite.
When sprite has no free space, than creates a new sprite with addition of a suffix to the sprite name.
Opera 9.0 and below have a bug which affects CSS background offsets less than -2042px. All values less than this are treated as -2042px exactly.

=item * B<max_sprite_height> [ = 2000 ]

Specify max height of sprite.

=item * B<margin> [ = 0 ]

Add margin to each image.

=item * B<transparent_color> [ = undef ]

Set transparent color for image, for example: 'white', 'black', ....

=item * B<is_background> [ = 0 ]

If B<is_background> flag is '0' will be generated CSS rule such as: 'width:XXXpx;height:YYYpx;'.

=item * B<layout> [ = 'packed' ]

Specify layout algorithm (horizontal, vertical or packed).

=item * B<css_selector_prefix> [ = '.spr-' ]

Specify CSS selector prefix.
For example, for an image "img/icon/arrow.gif" will be generated selector such as ".spr-img-icon-arrow".

=item * B<css_url_prefix> [ = '' ]

Specify prefix for CSS url.
For example: background-image: url('B<css_url_prefix>sample_sprite.png')

=item * B<is_add_timestamp> [ = 1 ]

If parameter set to '1' than timestamp will be added for CSS url.
For example: background-image: url('sample_sprite.png?12345678')

=back

=head2 build(<%args>)

Build sprites.

    $builder->build(<%args>);

This method returning structure like:

    {
        'sample_sprite_1.png' => [
            {
                'y' => 0,
                'width' => 32,
                'selector' => '.spr-small-add',
                'is_background' => 0,
                'x' => 0,
                'height' => 32,
                'image' => 'small/Add.png',
                'repeat' => 'no'
            },
            ...
        ],
        ...
    }

Valid arguments are:

=over

=item * B<sprites>

Specify sprite list.

    $builder->build(
        sprites => [
            {
                file               => 'horizontal_%d.png',
                layout             => 'horizontal',
                max_sprite_width   => 1000,
                ...
                images => [
                    { file => 'small/*.png', is_repeat => 1, ... },
                    { file => 'small/*.gif' },
                    { file => 'small/a.jpg', is_background => 1 },
                ],
            },
            {
                file                => 'sprite_%d.png',
                max_sprite_width    => 1000,
                css_selector_prefix => '#spr-',
                ...
                images => [
                    { file => 'small/Add.png', is_repeat => 1, ... },
                ],
            },
        ],
    );

=item * B<config>

Specify XML config filename (it requires XML::LibXML module).

    $builder->build(config => 'config.xml');

Example of config.xml:

    <root>
        <global_opts>
            <max_image_size>20000</max_image_size>
            <layout>packed</layout>
            <css_selector_prefix>.spr-</css_selector_prefix>
            <css_url_prefix>/sprite/</css_url_prefix>
        </global_opts>

        <sprites>
            <sprite file="sprite_%d.png">
                <image file="small/Add.png"/>
                <image file="small/Box.png"/>
                <image file="medium/CD.png"/>
            </sprite>
            <sprite file="sprite_x_%d.png" layout="vertical">
                <image file="small/Brick.png" is_repeat="1"/>
                <image file="small/Bin_Empty.png"/>
                <image file="medium/Close.png"/>
            </sprite>
            <sprite file="sprite_y_%d.png" layout="horizontal">
                <image file="small/Pattern.png" is_repeat="1"/>
                <image file="small/Address_Book.png"/>
                <image file="medium/Chat.png"/>
            </sprite>
        </sprites>
    </root>

=back

=head2 write_css([<filename|fh>])

Write CSS to file.
When B<filename> parameter is not specified than this method returning a string.

    $builder->write_css('sprite.css');

=head2 write_xml([<filename|fh>])

Write CSS sprites info structure into XML format (it requires XML::LibXML module).
When B<filename> parameter is not specified than this method returning a string.

    $builder->write_xml('sprite.xml');

Example of sprite.xml:

    <root>
        <sprite src="sample_sprite.png">
            <image y="0" width="32" selector=".spr-small-add" is_background="0" x="0" height="32" repeat="no" image="small/Add.png"/>
        </sprite>
    </root>


=head2 write_html([<filename|fh>])

Write HTML sample page.
When B<filename> parameter is not specified than this method returning a string.

=head2 get_sprites_info()

This method returning structure like:

    {
        'sample_sprite.png' => [
            {
                'y' => 0,
                'width' => 32,
                'selector' => '.spr-small-add',
                'is_background' => 0,
                'x' => 0,
                'height' => 32,
                'image' => 'small/Add.png',
                'repeat' => 'no'
            },
            ...
        ],
        ...
    }

=head1 AUTHOR

=over 4

Yuriy Ustushenko, E<lt>yoreek@yahoo.comE<gt>

=back

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013 Yuriy Ustushenko

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
